post-processing-code/general-functions/aggregation.py

```diff
@@ -0,0 +1,167 @@
+        pathToOutputFiles (str): the top level folder where the output files are located.
+        specific_files (str): file name filter to only combine the files that are wanted.
+        columns_to_replace_with_nan (list[str]): any columns that should be replaced with NaN rather than None.
+        typingMap (dict): a dictionary that maps column names to their dtype.
+    
+    Returns:
+        A dataframe with all of the CSVs stacked ontop of one another
+    """
+    rows = []
+    columns = []
+
+    files_to_combine = glob.glob(pathToOutputFiles + "**/" + specific_files, recursive=True)
+    index = 0
+    total_files = len(files_to_combine)
+    for filename in files_to_combine:
+        index += 1
+        with open(filename, newline="") as f:
+            reader = csv.reader(f)
+            if len(columns) == 0:
+                columns = next(reader)
+            else:
+                next(reader)
+            rows.extend(reader)
+        if ((total_files >= 10) and (index % (total_files // 10) == 0)):
+            print(f"Files Processed: {index / total_files * 100}%")
+    aggregatedData = pd.DataFrame(rows, columns=columns)
+    aggregatedData.loc[:, columns_to_replace_with_nan] = aggregatedData.loc[:, columns_to_replace_with_nan].replace("", np.nan)
+    aggregatedData.loc[:, ~aggregatedData.columns.isin(columns_to_replace_with_nan)] = aggregatedData.loc[:, ~aggregatedData.columns.isin(columns_to_replace_with_nan)].replace("", None)
+    aggregatedData = aggregatedData.astype(typingMap)
+    return aggregatedData
+
+def iuLevelAggregate(
+    df: pd.DataFrame,
+    filter_measures: list[str] = [],
+    measure_column_name: str = "measure"
+) -> pd.DataFrame:
+    """
+    Takes in an input dataframe of all the aggregated iu-lvl data and returns a filtered version if requested.
+
+    Args:
+        df (pd.Dataframe): the dataframe with all the iu-lvl data.
+        filter_measures (list[str]): a list contain the measures you want to filter. if empty (default), then it will not filter anything.
+        measure_column_name (list[str]): the name of the column where the measure names are located.
+    
+    Returns:
+        A dataframe with all of the iu-lvl data, filtered or not
+    """
+    if len(filter_measures) > 0:
+        df.loc[df[measure_column_name].isin(filter_measures), :]
+    return df
+
+def countryLevelAggregate(
+    df: pd.DataFrame,
+    measure_column_name: str = "measure",
+    general_summary_cols: list[str] = ["prevalence", "year_of_1_mfp_avg", "year_of_90_under_1_mfp"],
+    general_groupby_cols: list[str] = ["scenario", "country_code", "year_id", "age_start", "age_end", "measure"],
+    threshold_summary_cols: list[str] = ["year_of_1_mfp_avg", "year_of_90_under_1_mfp"],
+    threshold_groupby_cols: list[str] = ["scenario", "country_code", "measure"],
+    threshold_cols_rename: dict = {"year_of_90_under_1_mfp":"pct_of_ius_passing_90pct_threshold", "year_of_1_mfp_avg":"pct_of_ius_passing_avg_threshold"}
```

Comment:
💡 rather than having defaults it might make sense to have a file that contains some constant string and arrays that can be used, so that at the call site it is clearer what is being done, and easier to see how it can be tweaked.

Reply:
I like the defaults because you can see it clearly in the tooltips, your method does seem a bit clearer though. As a compromise I think I can not have them be defaulted parameters, but show the currently "default value" as an example in the function/args description